Implement power law estimation with 2 disconnect intervals, and `Offset` and `Polynomial` with arbitrary number of intervals

francisco-dlp opened the pull request:
### Description of the change

So far, `hs.model.compones1D.PowerLaw.estimate_parameters` could estimate the parameters of the power law using the two-area method on an interval defined by the `x1` and `x2`. Here we implement defining 2 disconnected intervals for estimation by adding the `x3` and `x4` keyword arguments. We also implement an arbitrary number of ROIs for `Offset` and `Polynomial` componets. 

### Progress of the PR
- [x] Change implemented (can be split into several points),
- [x] update docstring (if appropriate),
- [x] update user guide (if appropriate),
- [x] add an changelog entry in the `upcoming_changes` folder (see [`upcoming_changes/README.rst`](https://github.com/hyperspy/hyperspy/blob/RELEASE_next_minor/upcoming_changes/README.rst)),
- [x] Check formatting changelog entry in the `readthedocs` doc build of this PR (link in github checks)
- [x] add tests,
- [x] ready for review.

### Minimal example of the bug fix or the new feature
```python
import hyperspy.api as hs
import numpy as np

pl = hs.model.components1D.PowerLaw()
pl.r.value = 2
pl.A.value = 1e4
axis = np.arange(10, 20, 0.1)
s = hs.signals.Signal1D(pl.function(axis))
s.axes_manager[-1].scale = 0.1
s.axes_manager[-1].offset = 10
s.isig[15.:16.].data[:] = 0
s.add_poissonian_noise()
roi1 = hs.roi.SpanROI(11,14)
roi2 = hs.roi.SpanROI(17, 19)

pl.estimate_parameters(s,
                       intervals=[roi1, roi2],
                       only_current=True)

s2 = hs.signals.Signal1D(pl.function(axis))
s2.axes_manager[-1].scale = 0.1
s2.axes_manager[-1].offset = 10
cs = s + s2 * 1j
cs.plot()
roi1.add_widget(cs)
roi2.add_widget(cs)
```
![PL_two_intervals](https://github.com/hyperspy/hyperspy/assets/989719/d438bd48-77d6-4fc7-a201-aecd178fc933)

francisco-dlp committed:
Add possibility to perform a power law estimation in 2 discontinous regions

francisco-dlp committed:
Fix test

francisco-dlp committed:
Simplify the code

francisco-dlp committed:
Add sanity checks

francisco-dlp committed:
Split in 2 equal intervals for stability

francisco-dlp committed:
Use weigthed average for better accuracy

francisco-dlp committed:
Add tests

francisco-dlp committed:
Fix sanity check

francisco-dlp committed:
Another go at fixing it

francisco-dlp committed:
Fix test tolerance

francisco-dlp committed:
Add User Guide documentation

francisco-dlp committed:
Improve docstring

ericpre commented:
Is it possible to do more than 2 interval?
I am wondering, if we should think of a more generic API that would easily allow more than two internals, that could be used for other components?

jlaehne commented:
> Is it possible to do more than 2 interval? I am wondering, if we should think of a more generic API that would easily allow more than two internals, that could be used for other components?

I would also be interested to have two intervals, e.g. for background fitting of a linear function.

francisco-dlp commented:
Using the two-area method the function can only take 2 intervals. Eventually one could use the same function to estimate the background on more intervals and then find a way to consolidate the results.

Another approach could be linearizing the problem and using linear least squares to estimate the parameters. The we can have as many regions as we like. I have just tested this approach and the estimation is less robust to noise than the two-area method, and takes 2x more time for a single interval. However, it might be competitive timewise for more than 2 intervals.

@jlaehne, since the polynomial component uses `polyfit` to estimate the parameters, it should be possible to implement multiple-interval estimation for a linear function. 

I think that in this PR we could stick to the two-area method with 2 intervals. We can consider adding more intervals and/or adding multi-interval estimation to other components in a separate pull request. However, we should agree on the API for multiple intervals. Any suggestions?

ericpre commented:
ok, then we should implement an API that can support more than two intervals. The options could be:
1. Extend the API: `x1`/`x2` can be numbers or arrays, if arrays, `x1` and `x2` specify the start and the end of all intervals, respectively.
2. break the API with a deprecation: values need to be passed as an array of shape (n, 2), with n the number of interval (columns) and each row containing the start and end of a single interval.

Option 2 needs a bit more work to go through the deprecation but is possibly more convenient to use, as it avoid having to define start and end of a given interval in two different arrays. Another consideration is that this would consistent with the way the `integration_windows` are defined in https://hyperspy.org/exspy/reference/signals.html#exspy.signals.EDSSpectrum.get_lines_intensity. Off the top of my head, I can't of other functionalities that deals with parameters defining multiple intervals.

francisco-dlp commented:
Here are some alternatives:

- A `mask` keyword argument, which is consistent with how `Model` works.
- A `intervals` keyword argument that can take a list of tuples or ROIs defining the intervals.

ericpre commented:
> * A `intervals` keyword argument that can take a list of tuples or ROIs defining the intervals.

Yes, that would be convenient to support ROIs.

jlaehne commented:
> > * A `intervals` keyword argument that can take a list of tuples or ROIs defining the intervals.
> 
> Yes, that would be convenient to support ROIs.

I guess for using it with the background removal tool, ROIs would also make sense? But I have not looked into the inner workings of the function.

francisco-dlp committed:
Merge remote-tracking branch 'origin/RELEASE_next_minor' into ENH_two_region_PL_estimation

# Conflicts:
#	hyperspy/tests/component/test_components.py
#	hyperspy/tests/component/test_powerlaw.py

francisco-dlp committed:
Use VisibleDeprecationWarning for x1/x2/x3/x4 deprecation

Per HyperSpy conventions, use VisibleDeprecationWarning instead of
FutureWarning for API deprecation warnings.

Assisted-by: opencode:mimo-v2.5-pro

francisco-dlp committed:
Update User Guide documentation for intervals API

Update the power law estimation example to use the new intervals
parameter instead of the deprecated x1/x2/x3/x4 parameters.

Assisted-by: opencode:mimo-v2.5-pro

francisco-dlp committed:
Add intervals API for Polynomial.estimate_parameters

Support multiple intervals for polynomial fitting by concatenating
data from all intervals before polyfit. Backward compatible with
existing x1/x2 API.

Assisted-by: opencode:mimo-v2.5-pro

francisco-dlp committed:
Add  parameter to Offset.estimate_parameters

Accept list of tuples or SpanROI objects for multi-interval estimation.
Data from all intervals is concatenated before averaging.
Old x1/x2 API preserved with VisibleDeprecationWarning.

Assisted-by: opencode-go/mimo-v2.5-pro

francisco-dlp committed:
Add intervals parameter to Exponential.estimate_parameters

Accept list of tuples or SpanROI objects for multi-interval estimation.
Multiple intervals are merged into a single contiguous range.
Old x1/x2 API preserved with VisibleDeprecationWarning.

Assisted-by: opencode-go/mimo-v2.5-pro

francisco-dlp committed:
Add backward compat for positional only_current arg in intervals API

When callers pass only_current as a positional arg (old API), detect
bool value in intervals parameter and redirect correctly.

Assisted-by: opencode-go/mimo-v2.5-pro

francisco-dlp committed:
Revert Exponential intervals API - single interval doesn't justify new parameter

The Exponential estimation uses a two-half geometric mean method that
requires contiguous data. Since it can only accept one interval, the
existing x1/x2 API is sufficient.

Assisted-by: opencode-go/mimo-v2.5-pro

francisco-dlp committed:
Accept single SpanROI as intervals argument

Allow passing a single SpanROI object directly instead of wrapping
it in a list. This matches the SpanROI docstring: 'can be used in
place of a tuple'.

Assisted-by: opencode-go/mimo-v2.5-pro

francisco-dlp committed:
Remove single SpanROI support from intervals parameter

intervals now requires a list. Single SpanROI will be handled by
making x1/x2 accept ROIs directly in the future.

Assisted-by: opencode-go/mimo-v2.5-pro

francisco-dlp committed:
Fix Sphinx warnings: list of tuples -> list of tuple

Use 'tuple' (singular) in numpydoc type annotations to avoid
py:obj reference warnings during doc build.

Assisted-by: opencode-go/mimo-v2.5-pro

francisco-dlp committed:
Update User Guide and changelog for intervals API

- Document intervals parameter for PowerLaw, Polynomial, Offset
- Add Polynomial example to User Guide
- Update changelog entry for all three components

Assisted-by: opencode-go/mimo-v2.5-pro

francisco-dlp commented:
Implemented with the `intervals` keyword, now accepting arbitrary number of intervals for `Polynomial` and `Offset` componets (only 2 for `PowerLaw`).

francisco-dlp committed:
Fix version directive: merge v2.2 into v2.5

The two-disconnected-intervals feature is introduced in v2.5, not v2.2.

Assisted-by: opencode-go/mimo-v2.5-pro

francisco-dlp committed:
Move intervals parameter to end of signature

Remove backward compat workaround by placing intervals after
only_current in the parameter list.

Assisted-by: opencode-go/mimo-v2.5-pro

francisco-dlp committed:
Remove x3, x4 from PowerLaw signature

These parameters were added in this branch but are superseded by
intervals. Keep only x1, x2 for backward compatibility with the
original API.

Assisted-by: opencode-go/mimo-v2.5-pro

francisco-dlp committed:
Remove x1,x2 deprecation - they are standard API

x1, x2 are the standard way to pass ranges across all components.
intervals is an addition, not a replacement.

Assisted-by: opencode-go/mimo-v2.5-pro

francisco-dlp committed:
Remove x1,x2 deprecation from Offset

x1, x2 are standard API, not deprecated.

Assisted-by: opencode-go/mimo-v2.5-pro

francisco-dlp committed:
Address copilot review issues: intervals normalization, lazy handling, test coverage

- Fix intervals normalization: detect tuple-of-tuples vs bare pair in
  PowerLaw, Polynomial, Offset estimate_parameters
- Fix Polynomial lazy path: respect only_current in _get_concatenated_data
- Fix Offset lazy path: use da.concatenate instead of np.concatenate
- Add @lazifyTestClass to TestPowerLaw, TestOffset, TestPolynomial
- Restore gradient eval check for Expression components
- Add multi-interval tests for Offset and Polynomial
- Fix docs: import style and grammar markup
- Fix test setup: use map[values][:] + map[is_set][:] + fetch_stored_values

Assisted-by: opencode:mimo-v2.5-pro

francisco-dlp committed:
Restore default range when no intervals specified in PowerLaw.estimate_parameters

When x1/x2 are None and intervals is None, default to the full signal
range to preserve backward compatibility with remove_background and
other callers that don't pass explicit intervals.

Assisted-by: opencode:mimo-v2.5-pro

ericpre commented on hyperspy/_components/offset.py:
```diff
@@ -87,7 +99,50 @@ def estimate_parameters(self, signal, x1, x2, only_current=False):
+                elif isinstance(interval, (tuple, list)) and len(interval) == 2:
+                    interval_tuples.append(tuple(interval))
+                else:
+                    raise ValueError(
+                        f"Invalid interval format: {interval}. "
+                        "Expected tuple (left, right) or SpanROI object."
+                    )
+            indices = [axis.value_range_to_indices(a, b) for a, b in interval_tuples]
+        elif x1 is not None:
+            if x2 is None:
+                raise ValueError("x2 must be provided when using x1.")
+            i1, i2 = axis.value_range_to_indices(x1, x2)
+            indices = [(i1, i2)]
+        else:
+            indices = [(axis.low_index, axis.high_index + 1)]
```
Can this be made in a separate function to share with other components?

francisco-dlp replied:
Good idea, done!

ericpre reviewed (commented):
This looks good - it may just need a couple of tweaks for parsing the intervals, see my comment and remaining copilot's comments.

francisco-dlp committed:
Address review comments: full test coverage, extract shared _intervals_to_tuples

- Add comprehensive tests covering all intervals validation paths,
  lazy data branches, map=None paths, and out=True return path for
  PowerLaw, Offset, and Polynomial estimate_parameters methods.
- All three component modules now at 100% coverage.
- Extract duplicated intervals normalization into shared
  _intervals_to_tuples() in hyperspy/misc/model_tools.py (EricPre's
  suggestion).
- Remove dead code in Polynomial._get_concatenated_data (only_current
  was always True at the call site).
- Fix map_none tests to call estimate_parameters once first so
  superclass _estimate_parameters does not pre-emptively call
  _create_arrays.

Assisted-by: Claude:claude-sonnet-4-20250514

ericpre reviewed (commented):
There are failures for the build with oldest supported dependencies:
```python
FAILED hyperspy/tests/component/test_components.py::TestPowerLaw::test_lazy_estimate_parameters[True-True] - RuntimeError: Lazy support needs dask >= 2024.12.0
```

Either bump dask requirement or skip this test for older version of dask.

francisco-dlp committed:
Fix Sphinx warnings: unmatched  markers in PLOT2D_DOCSTRING

Two RST inline literal markers were missing closing backticks in the
PLOT2D_DOCSTRING template ('vmin' and 'auto' option in norm description),
causing 30 Sphinx warnings across all signal plot() methods that use this
shared docstring template.

These unmatched delimiters caused cascading RST parsing failures
(block quote ends without a blank line, unexpected indentation errors)
in complex_signal2d.py, signal2d.py, signal.py, complex_signal.py,
signal1d.py, lazy.py, model.py, model2d.py, and _artificial_data.py.

Assisted-by: deepseek:deepseek-v4-pro

francisco-dlp committed:
Merge remote-tracking branch 'origin/RELEASE_next_minor' into ENH_two_region_PL_estimation

francisco-dlp committed:
Add dask version skipif guards for lazy component tests

Tests using estimate_parameters in PowerLaw, Offset, and Polynomial
fail with old dask versions (< 2024.12.0). Skip those lazy tests
to keep CI green on older dependency matrices.

francisco-dlp committed:
Bump dask to >=2024.12.0 and fix polynomial lazy polyfit

- Bump dask minimum in pyproject.toml and conda_environment.yml
- Remove RuntimeError dead code and unused imports in model.py
- Remove version-based skipif guards in test_components.py
- Remove old-dask branches in test_model.py
- Make _skip_test a no-op in test_linear_model.py and test_remove_background.py
- Fix polynomial dead-code if/else in _get_concatenated_data
- Fix polynomial np.polyfit receiving dask arrays: compute before fitting
- Add changelog fragments

Assisted-by: Sisyphus:deepseek-v4-pro

ericpre committed:
Fix oldest supported version build
